controller/sharding/consistent/consistent.go

```diff
@@ -0,0 +1,288 @@
+// An implementation of Consistent Hashing and
+// Consistent Hashing With Bounded Loads.
+//
+// https://en.wikipedia.org/wiki/Consistent_hashing
+//
+// https://research.googleblog.com/2017/04/consistent-hashing-with-bounded-loads.html
+package consistent
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/btree"
+
+	blake2b "github.com/minio/blake2b-simd"
+)
+
+var ErrNoHosts = errors.New("no hosts added")
+
+type Host struct {
+	Name string
+	Load int64
+}
+
+type Consistent struct {
+	servers           map[uint64]string
+	clients           *btree.BTree
+	loadMap           map[string]*Host
+	totalLoad         int64
+	replicationFactor int
+
+	sync.RWMutex
+}
+
+type item struct {
+	value uint64
+}
+
+func (i item) Less(than btree.Item) bool {
+	return i.value < than.(item).value
+}
+
+func New() *Consistent {
+	return &Consistent{
+		servers:           map[uint64]string{},
+		clients:           btree.New(2),
+		loadMap:           map[string]*Host{},
+		replicationFactor: 1000,
+	}
+}
+
+func NewWithReplicationFactor(replicationFactor int) *Consistent {
+	return &Consistent{
+		servers:           map[uint64]string{},
+		clients:           btree.New(2),
+		loadMap:           map[string]*Host{},
+		replicationFactor: replicationFactor,
+	}
+}
+func (c *Consistent) Add(server string) {
+	c.Lock()
+	defer c.Unlock()
+
+	if _, ok := c.loadMap[server]; ok {
+		return
+	}
+
+	c.loadMap[server] = &Host{Name: server, Load: 0}
+	for i := 0; i < c.replicationFactor; i++ {
+		h := c.hash(fmt.Sprintf("%s%d", server, i))
+		c.servers[h] = server
+		c.clients.ReplaceOrInsert(item{h})
+	}
+}
+
+// Get returns the server that owns the given client.
+// As described in https://en.wikipedia.org/wiki/Consistent_hashing
+// It returns ErrNoHosts if the ring has no servers in it.
+func (c *Consistent) Get(client string) (string, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	if c.clients.Len() == 0 {
+		return "", ErrNoHosts
+	}
+
+	h := c.hash(client)
+	var foundItem btree.Item
+	c.clients.AscendGreaterOrEqual(item{h}, func(i btree.Item) bool {
+		foundItem = i
+		return false // stop the iteration
+	})
+
+	if foundItem == nil {
+		// If no host found, wrap around to the first one.
+		foundItem = c.clients.Min()
+	}
+
+	host := c.servers[foundItem.(item).value]
+
+	return host, nil
+}
+
+// GetLeast returns the least loaded host that can serve the key.
+// It uses Consistent Hashing With Bounded loads.
+// https://research.googleblog.com/2017/04/consistent-hashing-with-bounded-loads.html
+// It returns ErrNoHosts if the ring has no hosts in it.
+func (c *Consistent) GetLeast(client string) (string, error) {
```

Comment:
This function doesn't work with your implementation of the btree. At least not from the testing I did. Here are some modifications that work. Have tested with Argo CD end to end. Make modifications if needed:
```
func (c *Consistent) GetLeast(client string) (string, error) {
	c.RLock()
	defer c.RUnlock()

	if c.clients.Len() == 0 {
		return "", ErrNoHosts
	}
	h := c.hash(client)
	for {
		var foundItem btree.Item
		c.clients.AscendGreaterOrEqual(item{h}, func(bItem btree.Item) bool {
			if h != bItem.(item).value {
				foundItem = bItem
				return false // stop the iteration
			}
			return true
		})

		if foundItem == nil {
			// If no host found, wrap around to the first one.
			foundItem = c.clients.Min()
		}
		key := c.clients.Get(foundItem)
		if key != nil {
			host := c.servers[key.(item).value]
			if c.loadOK(host) {
				return host, nil
			}
			h = key.(item).value
		} else {
			return client, nil
		}
	}
}
```